File: lib/core/qr/qrPolynomial.js
goog.provide('qr.Polynomial');

/**
 @constructor
 */
qr.Polynomial = function(num, shift) {

  if (num.length == undefined) {
    throw new Error(num.length + '/' + shift);
  }

  var offset = 0;

  while (offset < num.length && num[offset] == 0) {
    offset++;
  }

  this.num = {
    length: num.length - offset + shift
  };
  for (var i = 0; i < num.length - offset; i++) {
    this.num[i] = num[i + offset];
  }
};

qr.Polynomial.prototype = {

  get: function(index) {
    return this.num[index];
  },

  getLength: function() {
    return this.num.length;
  },

  multiply: function(e) {

    var num = {
      length: this.getLength() + e.getLength() - 1
    };

    for (var i = 0; i < this.getLength(); i++) {
      for (var j = 0; j < e.getLength(); j++) {
        num[i + j] ^= qr.Math.gexp(qr.Math.glog(this.get(i)) + qr.Math.glog(e.get(j)));
      }
    }

    return new qr.Polynomial(num, 0);
  },

  mod: function(e) {

    if (this.getLength() - e.getLength() < 0) {
      return this;
    }

    var ratio = qr.Math.glog(this.get(0)) - qr.Math.glog(e.get(0));

    var num = {
      length: this.getLength()
    };

    var i;
    for (i = 0; i < this.getLength(); i++) {
      num[i] = this.get(i);
    }

    for (i = 0; i < e.getLength(); i++) {
      num[i] ^= qr.Math.gexp(qr.Math.glog(e.get(i)) + ratio);
    }

    // recursive call
    return new qr.Polynomial(num, 0).mod(e);
  }
};

File: lib/core/qr/qr_polynomial.dart
class QrPolynomial {
  List _myThings;

	QrPolynomial(List<int> thing, shift) {

    var offset = 0;

    while (offset < thing.length && thing[offset] == 0) {
      offset++;
    }

    _myThings = QrMath.getZeroedList(thing.length - offset + shift);

    for (var i = 0; i < thing.length - offset; i++) {
      _myThings[i] = thing[i + offset];
    }
  }

  operator [](int index) => _myThings[index];

  int get length() => _myThings.length;

  QrPolynomial multiply(QrPolynomial e) {

    List foo = QrMath.getZeroedList(length + e.length - 1);

    for (var i = 0; i < length; i++) {
      for (var j = 0; j < e.length; j++) {
        foo[i + j] ^= QrMath.gexp(QrMath.glog(this[i]) + QrMath.glog(e[j]));
      }
    }

    return new QrPolynomial(foo, 0);
  }

  QrPolynomial mod(e) {

    if (length - e.length < 0) {
      return this;
    }

    var ratio = QrMath.glog(this[0]) - QrMath.glog(e[0]);

    var thing = QrMath.getZeroedList(length);

    for (int i = 0; i < length; i++) {
      thing[i] = this[i];
    }

    for (int i = 0; i < e.getLength(); i++) {
      thing[i] ^= QrMath.gexp(QrMath.glog(e[i]) + ratio);
    }

    // recursive call
    return (new QrPolynomial(thing, 0)).mod(e);
  }
}
